crates/rolldown_common/src/chunk/mod.rs

```diff
@@ -114,14 +114,14 @@ impl Chunk {
     }
   }
 
-  pub fn has_side_effect(&self, runtime_id: ModuleIdx) -> bool {
-    // TODO: remove this special case, once `NormalModule#side_effect` is implemented. Runtime module should always not have side effect
-    if self.modules.len() == 1 && self.modules[0] == runtime_id {
-      return false;
-    }
-    // TODO: Whether a chunk has side effect is determined by whether it's module has side effect
-    // Now we just return `true`
-    true
+  pub fn has_side_effect(&self, module_table: &ModuleTable, runtime_id: ModuleIdx) -> bool {
+    self.modules.iter().any(|&module_id| {
+      // Runtime module should always not have side effect
+      if module_id == runtime_id {
+        return false;
+      }
+      module_table[module_id].side_effects().has_side_effects()
+    })
```

Comment:
```suggestion
    self.modules.iter().any(|&module_id| {
      module_table[module_id].side_effects().has_side_effects()
    })
```
Is it possible to simplify into this? It seems it has `DeterminedSideEffects::Analyzed(false)`.
https://github.com/rolldown/rolldown/blob/4c29202bfbc07b62ee0b166991ecf81c3a41a4f1/crates/rolldown/src/module_loader/runtime_module_task.rs#L131

Reply:
nice catch!